sql/bloom_filters.h

```diff
@@ -177,9 +184,153 @@ struct PatternedSimdBloomFilter
      basically, unnoticeable, well below the noise level */
 #endif
 
+#ifdef NEON_IMPLEMENTATION
+  uint64x2_t CalcHash(uint64x2_t vecData)
+  {
+    static constexpr uint64_t prime_mx2= 0x9FB21C651E98DF25ULL;
+    static constexpr uint64_t bitflip= 0xC73AB174C5ECD5A2ULL;
+    uint64x2_t step1= veorq_u64(vecData, vdupq_n_u64(bitflip));
+    uint64x2_t step2= veorq_u64(vshrq_n_u64(step1, 48), vshlq_n_u64(step1, 16));
+    uint64x2_t step3= veorq_u64(vshrq_n_u64(step1, 24), vshlq_n_u64(step1, 40));
+    uint64x2_t step4= veorq_u64(step1, veorq_u64(step2, step3));
+    uint64x2_t step5;
+    step5= vsetq_lane_u64(vgetq_lane_u64(step4, 0) * prime_mx2, step4, 0);
+    step5= vsetq_lane_u64(vgetq_lane_u64(step4, 1) * prime_mx2, step5, 1);
+    uint64x2_t step6= vshrq_n_u64(step5, 35);
+    uint64x2_t step7= vaddq_u64(step6, vdupq_n_u64(8));
+    uint64x2_t step8= veorq_u64(step5, step7);
+    uint64x2_t step9;
+    step9= vsetq_lane_u64(vgetq_lane_u64(step8, 0) * prime_mx2, step8, 0);
+    step9= vsetq_lane_u64(vgetq_lane_u64(step8, 1) * prime_mx2, step9, 1);
+    return veorq_u64(step9, vshrq_n_u64(step9, 28));
+  }
+
+  uint64x2_t GetBlockIdx(uint64x2_t vecHash)
+  {
+    uint64x2_t vecNumBlocksMask= vdupq_n_u64(num_blocks - 1);
+    uint64x2_t vecBlockIdx= vshrq_n_u64(vecHash, mask_idx_bits + rotate_bits);
+    return vandq_u64(vecBlockIdx, vecNumBlocksMask);
+  }
+
+  uint64x2_t ConstructMask(uint64x2_t vecHash)
+  {
+    uint64x2_t vecMaskIdxMask= vdupq_n_u64((1 << mask_idx_bits) - 1);
+    uint64x2_t vecMaskMask= vdupq_n_u64((1ull << bits_per_mask) - 1);
+
+    uint64x2_t vecMaskIdx= vandq_u64(vecHash, vecMaskIdxMask);
+    uint64x2_t vecMaskByteIdx= vshrq_n_u64(vecMaskIdx, 3);
+    int64x2_t vecMaskBitIdx=
+      vsubq_s64(vdupq_n_s64(0),
+                vreinterpretq_s64_u64(vandq_u64(vecMaskIdx, vdupq_n_u64(0x7))));
```

Comment:
could you explain this one, please?

Reply:
That's easy: it is equivalent to `-(int64_t) (maskIdx & 7)`, almost what default implementation does.

Negative is to turn left shift into right shift (`vshlq_u64(uint64x2_t a, int64x2_t b)`). There's no direct right shift call, it only supports shift by scalar: `vshrq_n_u64(uint64x2_t a, const int n)`.

Reply:
add a comment please, something like
```c
/* shift right in NEON is implemented as shift left by a negative value. do the negation here */
```